code/modules/pai/software.dm

```diff
@@ -0,0 +1,247 @@
+
+/mob/living/silicon/pai/ui_data(mob/user)
+	var/list/data = list()
+	data["door_jack"] = hacking_cable
+	data["image"] = card.emotion_icon
+	data["installed"] = installed_software
+	data["ram"] = ram
+	return data
+
+/mob/living/silicon/pai/ui_static_data(mob/user)
+	var/list/data = list()
+	data["available"] = available_software
+	data["directives"] = laws.supplied
+	data["emagged"] = emagged
+	data["languages"] = languages_granted
+	data["master_name"] = master_name
+	data["master_dna"] = master_dna
+	return data
+
+/mob/living/silicon/pai/ui_act(action, list/params, datum/tgui/ui)
+	. = ..()
+	if(.)
+		return
+	if(available_software[action] && !installed_software.Find(action))
+		to_chat(usr, span_warning("You do not have this software installed."))
+		CRASH("[usr] attempted to activate software they hadn't installed: [action]")
+	switch(action)
+		if("Atmospheric Sensor")
+			atmos_analyzer.attack_self(src)
+			return TRUE
+		if("buy")
+			buy_software(usr, params["selection"])
+			return TRUE
+		if("change image")
+			change_image(usr)
+			return TRUE
+		if("check dna")
+			check_dna(usr)
+			return TRUE
+		if("Crew Manifest")
+			ai_roster()
+			return TRUE
+		if("Crew Monitor")
+			GLOB.crewmonitor.show(usr, src)
+			return TRUE
+		if("Digital Messenger")
+			modularInterface?.interact(usr)
+			return TRUE
+		if("Door Jack")
+			// Look to door_jack.dm for implementation
+			door_jack(usr, params["mode"])
+			return TRUE
+		if("Encryption Slot")
+			to_chat(usr, span_notice("You have [!encrypt_mod ? "enabled" : "disabled"] encrypted radio frequencies."))
+			encrypt_mod = !encrypt_mod
+			radio.subspace_transmission = !radio.subspace_transmission
+			return TRUE
+		if("Host Scan")
+			host_scan(usr, params["mode"])
+			return TRUE
+		if("Internal GPS")
+			internal_gps.attack_self(src)
+			return TRUE
+		if("Music Synthesizer")
+			instrument.interact(src)
+			return TRUE
+		if("Medical HUD")
+			toggle_hud(usr, "medical")
+			return TRUE
+		if("Newscaster")
+			newscaster.ui_interact(src)
+			return TRUE
+		if("Photography Module")
+			// Look to pai_camera.dm for implementation
+			use_camera(usr, params["mode"])
+			return TRUE
+		if("Remote Signaler")
+			signaler.ui_interact(src)
+			return TRUE
+		if("Security HUD")
+			toggle_hud(usr, "security")
+			return TRUE
+		if("Universal Translator")
+			grant_languages(usr, ui)
+			return TRUE
+	return FALSE
+
+/**
+ * Purchases the selected software from the list and deducts their
+ * available ram.
+ *
+ * @param {mob} user - - The user purchasing the software.
+ * @param {string} selection - The software to purchase.
+ * @returns {boolean} - TRUE if the software was purchased, CRASH otherwise.
+ */
+/mob/living/silicon/pai/proc/buy_software(mob/user, selection)
+	if(!available_software[selection] || installed_software.Find(selection))
+		to_chat(user, span_warning("Error: Software unavailable."))
+		CRASH("[user] tried to purchase unavailable software as a pAI.")
```

Comment:
Remove all of these CRASH calls from anything that is a result of player input. Players inputting garbage data should not leave a runtime